### PR TITLE
feat: added homebrew package manager

### LINKED ISSUE #32: Homebrew Detection
**What module does this affect?**
`packages`

**What hardware/software isn't being detected right?**
Homebrew can be used on Linux not just macos

I personally use it to have pull my basic shell programs instead of layering rpm-ostree packages (makes updates sloooow) or exporting packages from a distrobox container (hassle). So using homebrew on linux is a thing.

**Does detection work a little, or is it completely broken?**
This does not work at all at the moment

### PR DESCRIPTION
**Pull request by FlyinPancake**
Closes #32

### CODE REVIEW

**Comment by LivacoNew** [comment]
Was it you on your IDE that caused all these styling changes? Either way, would prefer them to be left at least until I have time to write a style guide for this project. 

Also, have yet to install & try homebrew but would like to try to read the actual package data from the Cellar/Caskroom folders, would let version detection be able to query it for speed but will take a look at that myself later

**Comment by FlyinPancake** [comment]
Sorry, I didn't notice, I had format on save on. 

I basically looked at the fastfetch impl and then combined it with the flatpak impl, to get it working.

**Comment by LivacoNew** [comment]
Ah, I see. Ye FastFetch nor the Flatpak impl get package data, but would be prefered. Again can peep at this later as implementing it can be a bit of a pain

**Comment by FlyinPancake** [comment]
fixed the formatting :D

File: src/package_managers.rs [hunk 1]
```diff
@@ -250,6 +250,27 @@ impl ManagerInfo {
 
         Some(result as u64)
     }
+
+    pub fn process_homebrew_packages_count(&self) -> Option<u64> {
```
Comment (LivacoNew) [inline src/package_managers.rs:254]: Move this fn into the packages module itself, as it's only counting, not actually fetching package info.  I'll also fix this for flatpak later.

**Comment by FlyinPancake** [comment]
To be frank I don't really feel like reading the source code, but according to my research brew does not maintain a package db

**Review by LivacoNew — CHANGES_REQUESTED** [review]

**Comment by LivacoNew** [comment]
> To be frank I don't really feel like reading the source code, but according to my research brew does not maintain a package db

It should cus [you can query it apparently](https://docs.brew.sh/Querying-Brew)

File: src/package_managers.rs (hunk 1 above)
Comment (FlyinPancake) [reply src/package_managers.rs:254]: Are we aiming for std to be optional behind feature flags, or can I import read_dir on the top of the file?

**Comment by FlyinPancake** [comment]
Qerying takes a long ass time with their stuff

image

gonna dig in some more to see if they are just waking the dirs

**Comment by FlyinPancake** [comment]
The only thing I am sure about is that Ruby is slow as shit.

**Comment by LivacoNew** [comment]
I see they have basic name/version info, but nothing beyond that

image

Tried looking through brews source but had a seizure just looking at Ruby code

**Comment by LivacoNew** [comment]
Yeah, I'm happy with my idea of how to implement this now. Essentially go through the cellar/caskroom and every subdirectory is a package, from there each subdir is the version, which we can just sort alphabetically and select the last entry as at least most of my entries follow major/minor/patch or are just a date.

**Comment by FlyinPancake** [comment]
Welcome to the Rusticean's Ruby Support Group :laughing: 

In the pkgname/version dirs there is a hidden .brew dir, that houses a ruby file that contains the required metadata for ruby. It's a mess...

**Comment by LivacoNew** [comment]
> Welcome to the Rusticean's Ruby Support Group 😆

God help us all

Anyways, you wanna impl this or do you wanna just push this branch and I can do it? Up to you

**Comment by FlyinPancake** [comment]
I am happy to implement this

**Comment by FlyinPancake** [comment]
Well I just did some unspeakable crimes with mappnigs...

Added 2ms to the total run >.<

image

**Comment by LivacoNew** [comment]
> Added 2ms to the total run >.<

Not sure I like this, looking at your patch I *think* it's just cus you call a lot of re-loops. 
Mind testing with this patch, as I don't have enough packages on my system to make it a performance issue (only 400 microsecs): 
[hopefullyfaster.patch.txt](https://github.com/user-attachments/files/16923689/hopefullyfaster.patch.txt)

Just trying to reduce the amount of loops that get ran, although I understand that it's kinda unavoidable

**Comment by FlyinPancake** [comment]
Ran the patched version a couple of times:

image

**Comment by LivacoNew** [comment]
How many packages on your system?

**Comment by LivacoNew** [comment]
Also try to use a tool like hyperfine, as `time` can only go so far with it I've noticed before

**Comment by FlyinPancake** [comment]
Good thing I know just a tool to tell me how many packages do I have!

its 158

**Comment by LivacoNew** [comment]
In that case it's likely just difference in hardware, as i'm only seeing it need 427.911µs on both patches with 78 installed on brew

image

I also only just noticed the 20ms total run time on your screenshots, meaning ye likely just running slower. (Although do please test that FastFetch isn't faster than that, otherwise it's my fault with detections somehow and there's a problem lmao)

**Comment by FlyinPancake** [comment]
Done some tests with hyperfine

image

**Comment by LivacoNew** [comment]
Yeah it's fine then, I'll leave my patch to die then cus yours is a bit more concise

**Comment by FlyinPancake** [comment]
damn, I have a 16 core AMD 3950x and still

**Comment by FlyinPancake** [comment]
Could we rayon the shit out of the dirs?

**Review by LivacoNew — APPROVED** [review]
I'm iffy on the amount of `expect` calls that are used, I like good error checking, but it's minor and if it causes issues can just modify it later

**Comment by LivacoNew** [comment]
> Could we rayon the shit out of the dirs?

Don't wanna resort to multithreading just yet, lmao

**Comment by LivacoNew** [comment]
> damn, I have a 16 core AMD 3950x and still

Filesystem overhead then maybe? Not sure

**Comment by LivacoNew** [comment]
Right I'm happy to merge this now, thanks for the pr :)

**Comment by FlyinPancake** [comment]
> Filesystem overhead then maybe, if you have a sata drive and not a m.2 like I do? Not sure otherwise
I'm on an M2 ssd but maybe fedora and a bunch of bg tasks got in the way

> Don't wanna resort to multithreading just yet, lmao
Did some tests for my curiosity, the rayon overhead kills it sadly

> I'm iffy on the amount of expect calls that are used, I like good error checking, but it's minor and if it causes issues can just modify it later
I only used expects where I already checked the errors previously I believe

**Pull request merged**